### ISSUE TITLE
import error

### ISSUE THREAD

**Issue by 89z** [stretchr/testify#1113]
~~~
 github.com/walles/moar/m imports
        github.com/sirupsen/logrus tested by
        github.com/sirupsen/logrus.test imports
        github.com/stretchr/testify/assert loaded from github.com/stretchr/testify@v1.3.0,
        but go 1.16 would select v1.6.1
~~~

Example file:

~~~go
package hello

import (
   _ "github.com/89z/mech/musicbrainz"
   _ "github.com/walles/moar/m"
)
~~~

**Comment by brackendawson** [comment]
This appears to be related to https://github.com/walles/moar/issues/68

**Comment by 89z** [comment]
thats weird, I cant seem to repro the issue anymore. Maybe some package down the line updated. I dont know.

**Comment by 89z** [comment]
Reopening, because the issue is back for some reason. You can repro by trying to make a new module with this file:

~~~go
package m

import (
   _ "github.com/89z/mech/musicbrainz"
   _ "github.com/walles/moar"
)
~~~

**Comment by brackendawson** [comment]
That doesn't recreate for me, your example is invalid Go, it attempts to import main.

If you need help here can you expand on your issue with at least the very basics:
* Your environment, show us your `go env` output, what version of testify are you using?
* What you did, the actual command you ran.
* What happened.
* What you expected.
* Steps to reproduce, at least enough code to get the actual error.

**Comment by 89z** [comment]
> That doesn't recreate for me, your example is invalid Go, it attempts to import main.

Im not sure how to reply to this, as your comment doesnt even make sense. The example isnt invalid Go, it doesnt attempt to "import main", and finally, what does "import main" even mean? Im not a Go expert, but Ive been using Go for a while, and I have never even heard that phrase.

> * Your environment, show us your `go env` output

<details>

~~~
set GO111MODULE=
set GOARCH=amd64
set GOBIN=
set GOCACHE=C:\Users\Steven\AppData\Local\go-build
set GOENV=C:\Users\Steven\AppData\Roaming\go\env
set GOEXE=.exe
set GOEXPERIMENT=
set GOFLAGS=
set GOHOSTARCH=amd64
set GOHOSTOS=windows
set GOINSECURE=
set GOMODCACHE=C:\Users\Steven\go\pkg\mod
set GONOPROXY=
set GONOSUMDB=
set GOOS=windows
set GOPATH=C:\Users\Steven\go
set GOPRIVATE=
set GOPROXY=https://proxy.golang.org,direct
set GOROOT=c:\go
set GOSUMDB=sum.golang.org
set GOTMPDIR=
set GOTOOLDIR=c:\go\pkg\tool\windows_amd64
set GOVCS=
set GOVERSION=go1.17
set GCCGO=gccgo
set AR=ar
set CC=gcc
set CXX=g++
set CGO_ENABLED=1
set GOMOD=D:\Desktop\go.mod
set CGO_CFLAGS=-g -O2
set CGO_CPPFLAGS=
set CGO_CXXFLAGS=-g -O2
set CGO_FFLAGS=-g -O2
set CGO_LDFLAGS=-g -O2
set PKG_CONFIG=pkg-config
set GOGCCFLAGS=-m64 -mthreads -fmessage-length=0 -fdebug-prefix-map=C:\Windows\TEMP\go-build2962835219=/tmp/go-build -gno-record-gcc-switches
PS D:\Desktop>
~~~

</details>

> what version of testify are you using?

Just look at the original post, its there.

> * What you did, the actual command you ran.

~~~
go mod tidy
~~~

> * What happened.

Again, see original post.

> * What you expected.

No error

> * Steps to reproduce, at least enough code to get the actual error.

~~~
go mod tidy
~~~

**Comment by brackendawson** [comment]
OK, running `go mod tidy` with your example does produce a similar error, we did need to know that was what you were trying to do:
```
example.com/m imports
	github.com/walles/moar imports
	github.com/sirupsen/logrus tested by
	github.com/sirupsen/logrus.test imports
	github.com/stretchr/testify/assert loaded from github.com/stretchr/testify@v1.3.0,
	but go 1.16 would select v1.6.1
example.com/m imports
	github.com/walles/moar imports
	github.com/sirupsen/logrus tested by
	github.com/sirupsen/logrus.test imports
	github.com/stretchr/testify/require loaded from github.com/stretchr/testify@v1.3.0,
	but go 1.16 would select v1.6.1

To upgrade to the versions selected by go 1.16:
	go mod tidy -go=1.16 && go mod tidy -go=1.17
If reproducibility with go 1.16 is not needed:
	go mod tidy -compat=1.17
For other options, see:
	https://golang.org/doc/modules/pruning
```

However your example code is still invalid, `"github.com/walles/moar"` has the package name `main`, you cannot import `main`. Does this error reproduce using valid Go?

**Comment by 89z** [comment]
> However your example code is still invalid, `"github.com/walles/moar"` has the package name `main`, you cannot import `main`. Does this error reproduce using valid Go?

It doesnt matter that its invalid, as you dont need to compile the code to run `go mod tidy`. However since you are getting hung up on this, you can just use this instead:

~~~go
package hello

import (
   _ "github.com/89z/mech/musicbrainz"
   _ "github.com/walles/moar/m"
)
~~~

Same error.

**Comment by brackendawson** [comment]
I don't know if you've only pasted some of the error or if your Go installation produces a different error to mine (you deleted your reply with your go env?). But the error tells you what to do:

```
...
To upgrade to the versions selected by go 1.16:
	go mod tidy -go=1.16 && go mod tidy -go=1.17
If reproducibility with go 1.16 is not needed:
	go mod tidy -compat=1.17
For other options, see:
	https://golang.org/doc/modules/pruning
```

tl;dr run `go mod tidy -go=1.16 && go mod tidy -go=1.17`

**Comment by 89z** [comment]
> I don't know if you've only pasted some of the error or if your Go installation produces a different error to mine (you deleted your reply with your go env?). But the error tells you what to do:
> 
>
> ~~~
> To upgrade to the versions selected by go 1.16:
> 	go mod tidy -go=1.16 && go mod tidy -go=1.17
> If reproducibility with go 1.16 is not needed:
> 	go mod tidy -compat=1.17
> For other options, see:
> 	https://golang.org/doc/modules/pruning
> ~~~

I understand that, but this error only happens in some cases, which means its likely due to some module using an old Go version. So if the maintainers are willing to push the Go version, then the problem goes away without people having to run any special commands.

So unless you have commit privilege here, this issue isnt really addressed to you. I dont (and never did) need help troubleshooting, I need someone who is able to change the `go.mod` (or accept a PR).

**Comment by marten-seemann** [comment]
> To upgrade to the versions selected by go 1.16:
>	go mod tidy -go=1.16 && go mod tidy -go=1.17
> If reproducibility with go 1.16 is not needed:
>	go mod tidy -compat=1.17
> For other options, see:
>	https://golang.org/doc/modules/pruning

Is there any fix for this by now? I'm also seeing this, and it's making my CI build fail.

**Comment by jabielecki** [comment]
If for the thing* you are compiling you only support only Go 1.17 or above the answer is simple, change `go mod tidy` in your CI pipelines into `go mod tidy -compat=1.17`. This is like a declaration. You explicitly approve that your compiled binary will use different versions of dependencies than it used to use with Go 1.16. (The algorithm which chooses version of the dependencies changed with Go 1.17 in an incompatible way.)

If you ever compiled on 1.16, your program might have subtly different behavior, because it has different dependencies compiled in.

That's not the entire story, but this is sufficient solution for 99% of users.

[ * By the thing I mean _your_ program, not the `github.com/stretchr/testify` module itself. ]

**Comment by jabielecki** [comment]
Since testify can do nothing to prevent or fix the issue, **I'd recommend to close it**.

This error only happens when for an entire tree (!) of all (!) modules used in a given program, Go detects that the outcome differs between Go 1.16 algorithm and Go 1.17 algorithm. Since the algorithm deals mainly with diamond dependencies, and testify is a frequent diamond dependency, it will be often visible in these error messages. :man_shrugging:  Quick fix in  my previous comment. :point_up:

**Comment by 89z** [comment]
closing because I don't care anymore
